Kubectl: Plumb openapi validation (disabled by default)

apelisse opened the pull request:
**What this PR does / why we need it**: Creates a new flag '--openapi' and plumb in the validation code so that it can be used by default to validate objects against the openapi schema.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: partially https://github.com/kubernetes/kubectl/issues/49

**Special notes for your reviewer**:

This is not complete, the name of the variable must change for example.

**Release note**:
```release-note
Kubectl uses openapi for validation. If OpenAPI is not available on the server, it defaults back to the old Swagger.
```

Commit:
Validate against OpenAPI schema (if available)

Commit:
openapi: Use "group" to look for resources

Commit:
openapi: Handle properly empty/null fileds

Commit:
openapi-validation: Handle List special case

Commit:
openapi validation: Ignore unknown types

This follows the exact same logic as swagger.

apelisse commented:
@mengqiy have you seen the error? Do you know what's going on?

apelisse commented:
/retest

apelisse commented:
OK, We can't see the actual errors from the given link. I've reproduced locally, fixing ...

mengqiy commented:
>OK, We can't see the actual errors from the given link. I've reproduced locally, fixing ...

Not sure what link are you referring to.
Looks the tests are all passing. I assume you have fixed the issue.

apelisse commented:
Yeah, I have, feel free to take a look!

mengqiy commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -402,8 +404,20 @@ func (f *ring1Factory) AttachablePodForObject(object runtime.Object, timeout tim
 }
 
-func (f *ring1Factory) Validator(validate bool, cacheDir string) (validation.Schema, error) {
+func (f *ring1Factory) Validator(validate bool, openapi bool, cacheDir string) (validation.Schema, error) {
```
nit: `validate, openapi bool`

mengqiy commented on pkg/kubectl/cmd/util/openapi/validation/validation_test.go:
```diff
@@ -41,8 +43,30 @@ var _ = Describe("resource validation using OpenAPI Schema", func() {
 		Expect(validator).ToNot(BeNil())
 	})
 
+	It("finds and validates a deployment", func() {
```
> finds and validates a deployment

I don't understand what does "find" mean here.

mengqiy reviewed (commented):
2 nits

mengqiy commented:
/lgtm
Feel free to re-apply the label after addressing the nits :)

apelisse commented:
@mengqiy I've addressed your feedback (I think). I've also changed the name of the command-line argument (from `openapi` to `openapi-validation` to be more specific) and it's default from `true` to `false`.

I don't think it makes sense to have it enabled by default until #50404 gets merged. I'll make another PR to revert that value once the cache is "enabled".

mengqiy commented:
> I've also changed the name of the command-line argument (from openapi to openapi-validation to be more specific) and it's default from true to false.

I'm using `--openapi` for  kubectl apply as well.
But if you change the name, I can't reuse it in apply. I will need to add a separate flag.

WDYT if we have one flag gating all openapi related behavior?

apelisse commented:
> WDYT if we have one flag gating all openapi related behavior?

@pwittrock Do you have an opinion on that?

apelisse commented:
/retest

mengqiy commented:
/lgtm

apelisse commented:
@pwittrock can you please approve this?

pwittrock commented:
/approve

pwittrock commented:
/approve kubernetes/kubectl#49
